scripts: ci: check_compliance: Multiple performance improvements to check compliance script

kartben opened the pull request:
I am observing a ~50% speed improvement with this set of patches.
Still WIP

kartben committed:
scripts: ci:  check_compliance: cache git diff calls

We have an increasing number of tests and many of them end up making the
same git diff calls so let's cache them.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
scripts: ci: check_compliance: add parallel processing

Most compliance checks can be run in parallel, so add a command line
argument allowing to do so. Since some checks are not thread-safe
(typically the KConfig tests), also add the ability for them to
indicate that they don't support parallel execution.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

carlescufi commented:
Interesting, but is this worth the additional complexity? I am always cautious when it comes to these kind of changes. I have never seen compliance be a bottleneck anywhere, running locally or on CI.

kartben committed:
scripts: ci: check_compliance: speed up BoardYmlCheck

implement much more efficient parsing and half the time it takes to run
this otherwise rather long check

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
> Interesting, but is this worth the additional complexity? I am always cautious when it comes to these kind of changes. I have never seen compliance be a bottleneck anywhere, running locally or on CI. Note that I am mostly referring to the multiprocessing introduction.

Well, checks take over 20 seconds on a really fast machine which is actually starting to become a bit painful IMO, so I think it is worth trying to look at ways to optimize things.

pdgendt commented:
Running a profiler shows that `KconfigCheck` is the largest bottleneck (~%90)

```shell
$ python -m cProfile -o profile.out ./scripts/ci/check_compliance.py --annotate -c origin/main.. -e ClangFormat
$ snakeviz profile.out
```

carlescufi commented:
> > Interesting, but is this worth the additional complexity? I am always cautious when it comes to these kind of changes. I have never seen compliance be a bottleneck anywhere, running locally or on CI. Note that I am mostly referring to the multiprocessing introduction.
> 
> Well, checks take over 20 seconds on a really fast machine which is actually starting to become a bit painful IMO, so I think it is worth trying to look at ways to optimize things.

Makes sense, just weary about multiprocessing in a relatively simple script like this one.
